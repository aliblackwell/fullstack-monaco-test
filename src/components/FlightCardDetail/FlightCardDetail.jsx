import React from 'react';
import PropTypes from 'prop-types';
import STYLES from './FlightCardDetail.scss';
const getClassName = (className) => STYLES[className] || 'UNKNOWN';
import BpkText, { TEXT_STYLES } from 'bpk-component-text';
import LongArrowRightIconSm from 'bpk-component-icon/sm/long-arrow-right';
import BpkImage from 'bpk-component-image';

function getFlightTimeFromDateString(dateString) {
  const date = new Date(dateString);
  const timeWithDate = date.toLocaleDateString('en-GB', {
    hour: 'numeric',
    minute: '2-digit',
    hour12: false

  });
  return timeWithDate.split(', ')[1]; // just get the time
}

const LegDetails = ({ timestamp, airport }) => (
  <BpkText tagName="p" textStyle={TEXT_STYLES.base}>
    {getFlightTimeFromDateString(timestamp)}
    <br />
    <BpkText tagName="span" textStyle={TEXT_STYLES.base} className={getClassName('FlightCardDetail__airport')}>{airport}</BpkText>
  </BpkText>
);

LegDetails.propTypes = {
  timestamp: PropTypes.string,
  airport: PropTypes.string,
};

const FlightCardDetail = ({ flight }) => {
  const getFlightDuration = (flight) => {
    const duration = flight.duration_mins;
    return `${Math.floor(duration / 60)}h ${duration % 60}`;
  };

  const getFlightStops = (flight) => {
    const stops = flight.stops;
    return stops > 0 ? `${stops} Stop${stops > 1 ? 's' : ''}` : 'Direct';
  };

  return (
    <article className={getClassName('FlightCardDetail')}>
      <div className={getClassName('FlightCardDetail__leg-container')}>
        <div className={getClassName('FlightCardDetail__left')}>
          <BpkImage
            altText={`${flight.airline_name} logo`}
            style={{ height: '32px', width: '32px', minWidth: '32px' }}
            aspectRatio={32 / 32}
            src={`https://logos.skyscnr.com/images/airlines/favicon/${flight.airline_id}.png`}
          />
          <div className={getClassName('FlightCardDetail__times-container')}>
            <LegDetails timestamp={flight.departure_time} airport={flight.departure_airport} />
            <LongArrowRightIconSm aria-label="Arrow" className={getClassName('abc-icon__arrow')} />
            <LegDetails timestamp={flight.arrival_time} airport={flight.arrival_airport} />
          </div>
        </div>
        <div className={getClassName('FlightCardDetail__right')}>
          <BpkText
            className={getClassName('FlightCardDetail__flight-meta')}
            tagName="p"
            textStyle={TEXT_STYLES.sm}
          >
            {getFlightDuration(flight)}
            <br />
            <BpkText
              tagName="span"
              textStyle={TEXT_STYLES.sm}
              className={getClassName(
                `FlightCardDetail__flight-stops_${flight.stops > 0 ? 'indirect' : 'direct'}`
              )}
            >
              {getFlightStops(flight)}
            </BpkText>
          </BpkText>
        </div>
      </div>
    </article>
  );
};

FlightCardDetail.propTypes = {
  flight: PropTypes.object,
};

export default FlightCardDetail;
