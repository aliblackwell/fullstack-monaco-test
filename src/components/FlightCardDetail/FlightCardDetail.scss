@import '~bpk-mixins';

$secondaryText: rgb(118,112,132);

.FlightCardDetail {
  margin-bottom: $bpk-spacing-sm;
  width: 100%;
  p {
    margin: 0;
  }
  &__airport {
    color: $secondaryText;
  }
  &__left {
    display: flex; 
    width: 50%;     
  }
  &__right {
    width: 50%;
    text-align: right;
  }
  &__leg-container {
    display: flex; 
  }
  &__times-container {    
    margin-left: $bpk-spacing-sm;
    display: flex;
    width: calc(100% - 40px); // account for image + padding
    min-width: 120px;
    align-items: center;
    justify-content: space-between;
  }
  &__flight-meta {
    color: $secondaryText;
    line-height: 1.5 !important; // align baseline with flight time and airport
  }
  &__flight-stops_direct {
    color: darken(rgb(000, 166, 152), 7.5%) // Accessible Backpack Monteverde on white bg
  }
  &__flight-stops_indirect {
    color: rgb(203,59,81);
  }
}

.abc-icon__arrow {
  fill: rgb(169,164,181);
}
