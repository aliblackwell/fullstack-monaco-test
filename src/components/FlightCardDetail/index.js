import FlightCardDetail from './FlightCardDetail';
export default FlightCardDetail;
