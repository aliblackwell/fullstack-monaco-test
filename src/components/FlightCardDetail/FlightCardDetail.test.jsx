import React from 'react';
import TestRenderer from 'react-test-renderer';
import FlightCardDetail from './FlightCardDetail';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
const flight = {
  id: 'leg_1',
  departure_airport: 'BUD',
  arrival_airport: 'LTN',
  departure_time: '2020-10-31T15:35',
  arrival_time: '2020-10-31T17:00',
  stops: 0,
  airline_name: 'Wizz Air',
  airline_id: 'WZ',
  duration_mins: 145,
};
describe('FlightCardDetail', () => {
  it('should render correctly', () => {
    const tree = TestRenderer.create(<FlightCardDetail flight={flight} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display plural stops when stops > 1', () => {
    const customFlight = { ...flight };
    customFlight.stops = 2;
    const wrapper = mount(<FlightCardDetail flight={customFlight} />);
    expect(wrapper.text().includes('Stops')).toBe(true);
  });
  it('should display singular stop when stops < 2', () => {
    const customFlight = { ...flight };
    customFlight.stops = 1;
    const wrapper = mount(<FlightCardDetail flight={customFlight} />);
    expect(wrapper.text().includes('Stops')).toBe(false);
    expect(wrapper.text().includes('Direct')).toBe(false);
  });
  it('should display time in 24hr format', () => {
    const wrapper = mount(<FlightCardDetail flight={flight} />);
    expect(wrapper.text().includes('15:35')).toBe(true);
  });
  it("should show when it's a direct flight", () => {
    const wrapper = mount(<FlightCardDetail flight={flight} />);
    expect(wrapper.text().includes('Direct')).toBe(true);
  });
});
