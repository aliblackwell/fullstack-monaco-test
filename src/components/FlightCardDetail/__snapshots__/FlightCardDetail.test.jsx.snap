// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightCardDetail should render correctly 1`] = `
<article
  className="FlightCardDetail"
>
  <div
    className="FlightCardDetail__leg-container"
  >
    <div
      className="FlightCardDetail__left"
    >
      <div
        className={null}
        style={
          Object {
            "height": "32px",
            "minWidth": "32px",
            "width": "32px",
          }
        }
      >
        <div
          className="bpk-image bpk-image--no-background"
          style={
            Object {
              "height": 0,
              "paddingBottom": "100%",
            }
          }
          suppressHydrationWarning={false}
        >
          <img
            alt="Wizz Air logo"
            className="bpk-image__img"
            onLoad={[Function]}
            src="https://logos.skyscnr.com/images/airlines/favicon/WZ.png"
            suppressHydrationWarning={false}
          />
        </div>
      </div>
      <div
        className="FlightCardDetail__times-container"
      >
        <p
          className="bpk-text bpk-text--base"
        >
          15:35
          <br />
          <span
            className="bpk-text bpk-text--base FlightCardDetail__airport"
          >
            BUD
          </span>
        </p>
        <svg
          aria-label="Arrow"
          className="abc-icon__arrow"
          height="18"
          style={
            Object {
              "height": "1.125rem",
              "width": "1.125rem",
            }
          }
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M3 12a1.5 1.5 0 0 0 1.5 1.5h11.379l-4.94 4.94a1.5 1.5 0 0 0 2.122 2.12l7.5-7.5a1.5 1.5 0 0 0 0-2.12l-7.5-7.5a1.5 1.5 0 0 0-2.122 2.12l4.94 4.94H4.5A1.5 1.5 0 0 0 3 12z"
          />
        </svg>
        <p
          className="bpk-text bpk-text--base"
        >
          17:00
          <br />
          <span
            className="bpk-text bpk-text--base FlightCardDetail__airport"
          >
            LTN
          </span>
        </p>
      </div>
    </div>
    <div
      className="FlightCardDetail__right"
    >
      <p
        className="bpk-text bpk-text--sm FlightCardDetail__flight-meta"
      >
        2h 25
        <br />
        <span
          className="bpk-text bpk-text--sm FlightCardDetail__flight-stops_direct"
        >
          Direct
        </span>
      </p>
    </div>
  </div>
</article>
`;
